## Ill-conditioned problem
An ill-conditioned problem is a situation where the output of the problem is highly sensitive to small changes in the input. This characteristic is common in numerical computations such as solving equations or optimization problems. Ill-conditioning can lead to inaccurate results and numerical instability.

### Example of an Ill-Conditioned Problem

One classic example of an ill-conditioned problem is the solution of a system of linear equations, especially when the matrix of coefficients is close to singular. Here's a simple example:

Consider the system of linear equations given by the matrix equation $Ax = b$, where

$ A = \begin{bmatrix} 1 & 2 \\ 1.0001 & 2 \end{bmatrix}, \quad b = \begin{bmatrix} 3 \\ 3.0001 \end{bmatrix} $

Here, matrix $A$ is nearly singular because the rows of $A$ are nearly linearly dependent (they are very close to each other).



To further illustrate the problem of ill-conditioning in the matrix $A$ from the example above, we can see how a slight change in $A$ itself can affect the solution dramatically. We'll create a small perturbation in the matrix $A$ and observe how it impacts the solution $x$.

Here's what we'll do:
1. Perturb matrix $A$ slightly by adding a very small value to one of its elements.
2. Solve the linear system with the perturbed matrix.
3. Compare the solution with the original solution to see how much it changes.

Let's implement this to better understand the sensitivity caused by the ill-conditioning of $A$.

After perturbing matrix $A$ by adding just 0.0001 to its (2,2) entry, the new matrix $A$ becomes:

$ A_{\text{perturbed}} = \begin{bmatrix} 1 & 2 \\ 1.0001 & 2.0001 \end{bmatrix} $

For this perturbed matrix $A_{\text{perturbed}}$, the solution $x$ to the system $Ax = b$ changes to:

$ x = \begin{bmatrix} -1 \\ 2 \end{bmatrix} $

Compared to the original solution:

$ x = \begin{bmatrix} 1 \\ 1 \end{bmatrix} $

This change in the solution is substantial, given the minimal alteration to the matrix $A$. This dramatic change in the solution from a slight perturbation in the matrix elements underscores the ill-conditioned nature of the problem. Small inaccuracies in the data or computation can lead to significantly different results, making it crucial to handle such systems carefully, particularly in precision-sensitive applications.

In [2]:
import numpy as np

# Define the matrix A and vector b
A = np.array([[1, 2], [1.0001, 2]])
b = np.array([3, 3.0001])

# Compute the inverse of A if possible and the condition number
A_inv = np.linalg.inv(A)
cond_number = np.linalg.cond(A)

# Compute the solution x
x = np.linalg.solve(A, b)

# Create a small perturbation in b and solve again
b_perturbed = b + np.array([0.0001, 0])
x_perturbed = np.linalg.solve(A, b_perturbed)

cond_number, x, x_perturbed


(50001.000029925155, array([1., 1.]), array([0.     , 1.50005]))

### Finding Ill-Conditioning

To determine if a problem is ill-conditioned, you can:

1. **Condition Number**: Calculate the condition number of the matrix $A$ in the context of solving $Ax = b$. The condition number, given by $\kappa(A) = \|A\| \|A^{-1}\|$, measures the sensitivity of the function's output relative to its input. A high condition number (significantly greater than 1) indicates an ill-conditioned matrix.

2. **Eigenvalues**: Analyze the eigenvalues of $A$. If the ratio of the largest to smallest eigenvalue (absolute values) is large, the matrix is ill-conditioned.

3. **Numerical Experiments**: Introduce small perturbations to $A$ and $b$ and observe the variation in the solution $x$. Large variations suggest ill-conditioning.



### Solving an Ill-Conditioned Problem

To solve or mitigate the effects of an ill-conditioned problem, you can:

1. **Improve Numerical Precision**: Use higher precision in your calculations if possible. This reduces the rounding errors that exacerbate ill-conditioning effects.

2. **Regularization**: Add a small value, say $\epsilon$, to the diagonal elements of $A$ (if applicable). This technique, known as regularization, can help make $A$ more stable and less sensitive to perturbations in $b$.

3. **Use Stable Algorithms**: Employ numerical methods that are specifically designed to handle ill-conditioned matrices. For example, using the Singular Value Decomposition (SVD) for solving linear equations is more stable than using the normal equation directly.

4. **Rescaling**: Rescale the problem when possible to balance out the scale of the data, which can improve the condition number of the matrix.

Let's demonstrate how to calculate the condition number of matrix $A$ using Python, and then perturb $b$ slightly to observe the sensitivity in the solution.

The condition number of the matrix $A$ is approximately 50,001, which is very high, indicating that the problem is indeed ill-conditioned.

For the original vector $b$, the solution $x$ to the system $Ax = b$ is:

$ x = \begin{bmatrix} 1 \\ 1 \end{bmatrix} $

However, after introducing a tiny perturbation to $b$ (adding 0.0001 to the first component), the solution $x$ changes drastically to:

$ x = \begin{bmatrix} 0 \\ 1.50005 \end{bmatrix} $

This significant change in the solution from a minor change in $b$ exemplifies the sensitivity due to the ill-conditioning of the matrix $A$.

To address this problem, one might consider techniques like regularization or using more stable numerical methods, such as SVD, as mentioned earlier. These approaches can help reduce the impact of such sensitivity and produce more reliable results.